Add support for OV5640 camera

jasaw opened the pull request:
Add support for OV5640 5MP camera. The camera has a resolution of 2560 x 1920 but the output will be scaled down to the configured resolution e.g. 640 x 480. When zoom is enabled, a 640 x 480 window is cropped from the full 2560 x 1920 frame, therefore giving double the magnification compared to OV2640.

This is a zoom comparison between OV5640 and OV2640.

OV5640:
![ov5640](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/dea3f8b2-83b4-412d-9046-9d4aa85037ad)

OV2640:
![ov2640](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/1e2c3c6f-e532-48e5-96f1-be66a6b1fdf1)

caco3 commented:
what is the impact on the RAM usage?
Does it need to store a larger framebuffer for the higher resolution?

jasaw commented:
@caco3 There's no impact on RAM usage at all. It works the same way as ov2640 where cropping is done on the camera module itself.

SybexX commented:
Process:
1. The camera takes an image with a resolution of 2560x1920
2. The camera cuts out the desired area from the image
3. The camera scales the cropped area to 640x480, if needed (desired area is larger 640x480)
This process applies both when using the zoom function and when zoom is deactivated.

However, further changes to the implementation are necessary, e.g. at the sharpness....


Ablauf:
1. Die Kamera nimmt ein Bild mit einer Auflösung von 2560x1920 auf
2. Die Kamera schneidet den gewünschten Bereich aus dem Bild aus
3. Die Kamera skaliert den ausgeschnittenen Bereich bei Bedarf auf 640x480 (wenn der ausgeschnittene Bereich größer 640x480 ist).
Dieser Ablauf wird sowohl bei der Nutzung der Zoom-Funktion als auch bei deaktivierten Zoom benutzt/verwendet.

Allerdings sind noch weitere Änderungen an der Implementierung notwendig, z.B. bei sharpness.....

jasaw commented:
>However, further changes to the implementation are necessary

@SybexX May I know what you would like to change?

SybexX commented:
e.g. so:
![sharpness](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/7231b71c-9505-4958-930c-17a6a13c5f08)

I made a few quick adjustments, but I don't know if it works because I don't have an OV5640.
[ClassControllCamera.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15231764/ClassControllCamera.zip)

jasaw commented:
@SybexX I've adjusted the sharpness handling to your version and it's working fine on my OV5640 system. I changed the sharpness range because the officially supported sharpness range is from -3 to +3.
`_sharpnessLevel = min(2, max(-2, _sharpnessLevel));`

Your `ClassControllCamera` code enforces 4:3 ratio on the zoom parameters, but OV5640 is a lot more flexible. It doesn't have the 4:3 restriction. Currently this PR only supports zoom mode 0, but it might be possible to achieve other zoom modes by passing different parameters to `set_res_raw` with scaling and binning set to `true`. That needs more experimentation, which I think could be left for future enhancement.

SybexX commented:
I specifically set the sharpnessLevel to 2. 
After longer tests, I found that a value of 3 sometimes causes problems.

I know that the OV5640 can do more, but we should adapt everything to the OV2640, 
otherwise we would need at least two pages, i.e. one for each camera model and it simply becomes too confusing for the user.

Sometimes less is more^^

SybexX commented:
![kamera](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/04b85f60-5c3b-40ab-97b3-d94b93598618)

jasaw commented:
@SybexX Thank you for reviewing. I've refactored my code based on your sample code but had to make a few changes to make it work with the OV5640.

> only else, otherwise the other supported camera models will be ignored

My apologies, it is not clear to me that OV3660 (which is the other supported camera) behaves the same way as OV2640 and parts of the code are hitting the OV2640 registers directly (presumably the OV3660 has the same registers as OV2640?). Personally, I prefer to be more explicit in the code when it comes to handling hardware specific behaviour, like
```
if (sensor_info->model == CAMERA_OV5640)
{
}
else if ((sensor_info->model == CAMERA_OV2640) || (sensor_info->model == CAMERA_OV3660))
{
}
```
Either way, I am not particular with how the code is written as long as it is well documented via comments or code itself so other developers understand the intention of the author.

SybexX commented:
The registers don't matter, the properties of the camera are more important.
The https://github.com/espressif/esp32-camera relieves us of having to pay attention to the registers.
Since no distinction was made between the camera models before I made my changes, I assumed that the OV2640 and OV3660 were very similar. After I quick look, the OV3660 is very close to the OV5640 in terms of functions.

jasaw commented:
>The registers don't matter

`ov2640_set_sharpness` and `ov2640_enable_auto_sharpness` functions are called for OV3660 camera as well and eventually those functions call `sensor->set_reg` which writes directly to the OV3660's registers. Keep in mind that sharpness is not officially supported by Espressif SDK. I haven't looked into the OV3660 datasheet, so I don't know what the camera will do if those registers don't exist or worse, have a different function. I don't have an OV3660 to test, so can't validate.

SybexX commented:
![sharpness](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/f44fa10a-4ea9-4f60-831a-ae089a54ceda)

jasaw commented:
I propose this sharpness handling code:
```cpp
void CCamera::SetCamSharpness(bool _autoSharpnessEnabled, int _sharpnessLevel)
{
    sensor_t *s = esp_camera_sensor_get();

    if (s != NULL)
    {
        _sharpnessLevel = min(2, max(-2, _sharpnessLevel));

        camera_sensor_info_t *sensor_info = esp_camera_sensor_get_info(&(s->id));

        if (sensor_info != NULL)
        {
            if (sensor_info->model == CAMERA_OV5640 || sensor_info->model == CAMERA_OV3660)
            {
                if (_autoSharpnessEnabled)
                {
                    // autoSharpness is not supported, default to zero
                    s->set_sharpness(s, 0);
                }
                else
                {
                    s->set_sharpness(s, _sharpnessLevel);
                }
            }
            else if (sensor_info->model == CAMERA_OV2640)
            {
                // The OV2640 does not officially support sharpness, so the detour is made with the ov2640_sharpness.cpp.
                if (_autoSharpnessEnabled)
                {
                    s->set_sharpness(s, 0);
                    ov2640_enable_auto_sharpness(s);
                }
                else
                {
                    ov2640_set_sharpness(s, _sharpnessLevel);
                }
            }
        }
    }
    else
    {
        LogFile.WriteToFile(ESP_LOG_ERROR, TAG, "SetCamSharpness, Failed to get Cam control structure");
    }
}

```

jasaw commented:
@SybexX Looks like OV3660 `set_res_raw` function is the same as OV5640 as well. I should probably handle OV3660 explicitly.

SybexX commented:
or so, then the ESP is spared the second comparison ^^

void CCamera::SetCamSharpness(bool _autoSharpnessEnabled, int _sharpnessLevel)
{
    sensor_t *s = esp_camera_sensor_get();

    if (s != NULL)
    {
        _sharpnessLevel = min(2, max(-2, _sharpnessLevel));

        camera_sensor_info_t *sensor_info = esp_camera_sensor_get_info(&(s->id));

        if (sensor_info != NULL)
        {
            if (sensor_info->model == CAMERA_OV2640)
            {
                // The OV2640 does not officially support sharpness, so the detour is made with the ov2640_sharpness.cpp.
                if (_autoSharpnessEnabled)
                {
                    s->set_sharpness(s, 0);
                    ov2640_enable_auto_sharpness(s);
                }
                else
                {
                    ov2640_set_sharpness(s, _sharpnessLevel);
                }
            }
            else
            {
				// for CAMERA_OV5640 and CAMERA_OV3660
                if (_autoSharpnessEnabled)
                {
                    // autoSharpness is not supported, default to zero
                    s->set_sharpness(s, 0);
                }
                else
                {
                    s->set_sharpness(s, _sharpnessLevel);
                }
            }			
        }
    }
    else
    {
        LogFile.WriteToFile(ESP_LOG_ERROR, TAG, "SetCamSharpness, Failed to get Cam control structure");
    }
}

SybexX commented:
if (sensor_info->model == CAMERA_OV2640)
            {
                // The OV2640 does not officially support sharpness, so the detour is made with the ov2640_sharpness.cpp.
                if (_autoSharpnessEnabled)
                {
                    s->set_sharpness(s, 0);	<<<<<<<<<< oh yes, this is actually not necessary, as it only returns a -1(not supported)
                    ov2640_enable_auto_sharpness(s);
                }
                else
                {
                    ov2640_set_sharpness(s, _sharpnessLevel);
                }
            }

jasaw commented:
@SybexX I have updated the code to your suggested sharpness handling and also updated zoom handling code to better deal with OV3660 but I don't have an OV3660 so can't test. I have only tested with OV5640 so far.

SybexX commented:
case CAMERA_OV5640:
                    frameSizeX = 2560;
                    frameSizeY = 1920;
                    // max imageSize = ((frameSizeX - CCstatus.ImageWidth) / 8 / 4) - 1
                    // 59 = ((2560 - 640) / 8 / 4) - 1
                    if (imageSize < 59)
                    {
                        _imageSize_temp = (59 - imageSize);
                    }
                    SanitizeZoomParams(_imageSize_temp, frameSizeX, frameSizeY, _imageWidth, _imageHeight, _offsetx, _offsety);
                    SetCamWindow(s, unused, frameSizeX, frameSizeY, _offsetx, _offsety, _imageWidth, _imageHeight, CCstatus.ImageWidth, CCstatus.ImageHeight);
                    break;

                case CAMERA_OV3660:
                    frameSizeX = 2048;
                    frameSizeY = 1536;
                    // max imageSize = ((frameSizeX - CCstatus.ImageWidth) / 8 / 4) -1
                    // 43 = ((2048 - 640) / 8 / 4) - 1
                    if (imageSize < 43)
                    {
                        _imageSize_temp = (43 - imageSize);
                    }
                    SanitizeZoomParams(_imageSize_temp, frameSizeX, frameSizeY, _imageWidth, _imageHeight, _offsetx, _offsety);
                    SetCamWindow(s, unused, frameSizeX, frameSizeY, _offsetx, _offsety, _imageWidth, _imageHeight, CCstatus.ImageWidth, CCstatus.ImageHeight);
                    break;

                case CAMERA_OV2640:
                    // ov2640_sensor_mode_t _mode = OV2640_MODE_UXGA; // 1600x1200
                    // ov2640_sensor_mode_t _mode = OV2640_MODE_SVGA; // 800x600
                    // ov2640_sensor_mode_t _mode = OV2640_MODE_CIF;  // 400x296
                    _mode = 0;
                    frameSizeX = 1600;
                    frameSizeY = 1200;
                    // max imageSize = ((frameSizeX - CCstatus.ImageWidth) / 8 / 4) -1
                    // 29 = ((1600 - 640) / 8 / 4) - 1
                    if (imageSize < 29)
                    {
                        _imageSize_temp = (29 - imageSize);
                    }
                    SanitizeZoomParams(_imageSize_temp, frameSizeX, frameSizeY, _imageWidth, _imageHeight, _offsetx, _offsety);
                    // _mode sets the sensor resolution (3 options available),
                    // _offsetx and _offsety set the start of the ROI,
                    // _imageWidth and _imageHeight set the size of the ROI,
                    // CCstatus.ImageWidth and CCstatus.ImageHeight set the output window size.
                    SetCamWindow(s, _mode, frameSizeX, frameSizeY, _offsetx, _offsety, _imageWidth, _imageHeight, CCstatus.ImageWidth, CCstatus.ImageHeight);
                    break;

jasaw commented:
@SybexX The new `imageSize` code that you shared above appears to be working well on my OV5640 setup and thanks for adding the comments to explain how the `imageSize` numbers came from because I had no idea previously.

I've also increased the OV5640 full frame size to match the datasheet.

SybexX commented:
still necessary adjustment in the edit_reference.html:

            <td>
                    <input required type="number" id="TakeImage_CamZoomSize_value1" value="0" min="0" max="59" step="1" onchange="cameraParameterChanged()"
                    oninput="(!validity.rangeOverflow||(value=59)) && (!validity.rangeUnderflow||(value=0)) && (!validity.stepMismatch||(value=parseInt(this.value)));">
            </td>


still necessary adjustment in the edit_config_template.html:
			<td>
				<input required type="number" id="TakeImage_CamZoomSize_value1" value="0" min="0" max="59" step="1" onchange="cameraParameterChanged()"
				oninput="(!validity.rangeOverflow||(value=59)) && (!validity.rangeUnderflow||(value=0)) && (!validity.stepMismatch||(value=parseInt(this.value)));">
			</td>
			<td>$TOOLTIP_TakeImage_CamZoomSize</td>			
		</tr>

		<tr class="expert" unused_id="TakeImage_CamZoomOffsetX_ex3">
			<td class="indent1">
				<class id="TakeImage_CamZoomOffsetX_text" style="color:black;">Zoom Offset X</class>
			</td>
			<td>
				<input required type="number" id="TakeImage_CamZoomOffsetX_value1" value="0" min="-960" max="960" step="8" onchange="cameraParameterChanged()"
				oninput="(!validity.rangeOverflow||(value=960)) && (!validity.rangeUnderflow||(value=-960)) && (!validity.stepMismatch||(value=parseInt(this.value)));">Pixel
			</td>
			<td>$TOOLTIP_TakeImage_CamZoomOffsetX</td>
		</tr>

		<tr class="expert" unused_id="TakeImage_CamZoomOffsetY_ex3">
			<td class="indent1">
				<class id="TakeImage_CamZoomOffsetY_text" style="color:black;">Zoom Offset Y</class>
			</td>
			<td>
				<input required type="number" id="TakeImage_CamZoomOffsetY_value1" value="0" min="-720" max="720" step="8" onchange="cameraParameterChanged()"
				oninput="(!validity.rangeOverflow||(value=720)) && (!validity.rangeUnderflow||(value=-720)) && (!validity.stepMismatch||(value=parseInt(this.value)));">Pixel
			</td>
			<td>$TOOLTIP_TakeImage_CamZoomOffsetY</td>
		</tr>

SybexX commented:
@jasaw Have you already tested it with an OV2640?
For me it always comes to a reboot at take_image, there must be something wrong with the calculation.

SybexX commented:
I made a few changes so that the OV2640 works again.  [test.txt](https://github.com/jomjol/AI-on-the-edge-device/files/15284481/test.txt)

jasaw commented:
@SybexX Thank you for testing with OV2640. I have been testing with OV5640 and haven't got time to test against OV2640.

I have incorporated your recommended changes. Thanks for fixing it and adding comments to explain how `imageSize` is supposed to work. I have added your commented in the code to help other developers.

I have tested your multi-step zoom and it is a very nice idea. I have to use `scale` and `binning` for OV5640 to work. I have also been testing the OV5640 in less than perfect lighting conditions and have to bump the max jpeg quality down to 18 because the camera is unstable below 18. I don't have an OV3660 so I can only assume it works the same way as OV5640.  I noticed you changed the max jpeg quality to 6 for OV2640. Is the OV2640 stable at 6 in less than perfect lightning conditions?

SybexX commented:
I always test on the water meter and there is no problem with the quality = 6.
Have you tried increasing the frequency, it is set to 20MHz, but I think the OV5640 can go up to 27MHz

jasaw commented:
>Have you tried increasing the frequency, it is set to 20MHz, but I think the OV5640 can go up to 27MHz
The OV5640 datasheet recommends 24MHz and I tried that but it didn't work properly. The image was always half black with some ghosting and the other half of the image was normal.

jasaw commented:
@SybexX I've checked other examples of running OV5640 on ESP32 and they are running 20MHz clock from what I can see. Not sure what's the maximum the ESP32 + OV5640 combo can handle.

I have just found out that both OV3660 and OV5640 are different from OV2640 in these areas as well:
1. OV3660 and OV5640 support de-noise. The range is from 0 to 8 where zero is auto-denoise. OV2640 does not support de-noise.
2. OV3660 and OV5640 auto-exposure range is from -5 to +5 whereas OV2640 range is from 1 to 5.
3. OV3660 and OV5640 sharpness range is from from -3 to +3 but no auto-sharpness whereas we limit our OV2640 range to -2 to +2 and `ov2640_set_sharpness` takes -3 to +3.
What would you like to do? I'm leaning towards increasing the sharpness range to -3 to +3, increasing the auto-exposure range to -5 to +5 but clip it internally for OV2640, and add de-noise option on the web interface.

SybexX commented:
Personally, I would limit myself to the functions of the OV2640.
Users won't change most of the settings anyway if the image quality is right.
That's why I only put the most necessary settings on the reference page and the rest is only available on the settings page.
It is best if @jomjol or @caco3 comments on what and how he imagines the implementation of the available settings and their setting range.

friedpa commented:
I personnaly think that the AutoSharpness (if that is the same as AutoFocus) is most important, as a lot of users have difficulties with the manual lens adjustment. So it could be better to change to the new camera as a default hardware.

SybexX commented:
AutoSharpness and AutoFocus are two different functions.
The problem is that every auto function can change the image quality every time and this making the evaluation a game of chance.
For a reasonable and consistent evaluation, the image quality must always remain the same.

jasaw commented:
For my gas meter and water meter set up, I can't have a fully enclosed housing because meter readers come around every couple of months to read the meters and they need to be able to look at the digits on the meters. What this means is my image is highly subjected to ambient lighting (sun light), so auto-exposure needs to be on. Other auto functions may need to be turned on to get an optimal image.

Regarding the focus issue, I agree that it's hard to get the focus right, which is why I have been researching the OV5640 with autofocus. I think it will be a more user friendly solution. The user adjusts the focus through the web interface during edit-reference step and leave the focus at the configured level during normal operation (no auto-focus). We would need to proof that focus can be adjusted manually first. I've been sharing my research here: https://github.com/jomjol/AI-on-the-edge-device/issues/2162#issuecomment-2102013159

SybexX commented:
I adjusted a few things and also added Denoise(Config-Page), all the changed files are in the zip.
However, I haven't tested it yet^^
[AI-on-the-edge-device_ov5640.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15296803/AI-on-the-edge-device_ov5640.zip)

jasaw commented:
@SybexX  I've incorporated your changes and cleaned the code up a little. My OV5640 is still working, which is good. The higher level of auto-exposure is definitely having an impact on the image. I haven't got time to test de-noise yet and how much impact it has on image quality.

SybexX commented:
@jasaw I bought an OV5640 for testing and found that the colors red and blue swapped in the zoom function^^
I also had this problem with the OV2640 and only noticed it when I tested it in my apartment and not on the water meter^^

With this change they are back to normal for me:
s->set_res_raw(s, xOffset, yOffset, xOffset + xTotal + 1, yOffset + yTotal + 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning);

Does your camera get so warm that the temperature of the ESP rises by around 6 to 10 degrees?
With the OV2640 I had around 34 degrees, now it's around 43 degrees^^

jasaw commented:
@SybexX I don't have swapped red blue colour problem on both my OV5640. I don't understand how `s->set_res_raw(s, xOffset, yOffset, xOffset + xTotal + 1, yOffset + yTotal + 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning)` would make any difference.

The OV5640 does run a little warmer than the OV2640, so I stuck the camera against the microSD card holder using a thermal tape, using the ESP's PCB as a heatsink. I have 2 OV5640 systems, one of them running at 85 degrees (under direct sunlight), and another running at 37 degrees (shaded). I have just purchased a few heatsinks to put on the ESP32 metal can.

Could your camera red blue colour swap problem be caused by high temperature?

SybexX commented:
It won't be the temperature, maybe the version of the camera, because I can set the quality to 8 on this camera without any problems (in very poor lighting conditions).
With the standard settings I find the image overall to be significantly darker compared to the ov2640.
xOffset + xTotal + 1, yOffset + yTotal + 1 increase the ISP input size and therefore it is no longer divisible by 8 without remainder.

jasaw commented:
@SybexX I can't get the `xOffset + xTotal + 1, yOffset + yTotal + 1` to work with my OV5640, the image is always corrupted. What happens when you do `s->set_res_raw(s, xOffset + 1, yOffset + 1, xOffset + xTotal + 1, yOffset + yTotal + 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning);` ? Does it fix the colour swap issue?

This is my OV5640 as reported by the software:
`MAIN: Camera info: PID: 0x5640, VER: 0x00, MIDL: 0x00, MIDH: 0x00`

SybexX commented:
s->set_res_raw(s, xOffset + 1, yOffset + 1, xOffset + xTotal + 1, yOffset + yTotal + 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning); works too.

I (4955) cam_hal: cam config ok
I (4995) ov5640: Set PLL: bypass: 0, multiplier: 180, sys_div: 4, pre_div: 2, root_2x: 0, pclk_root_div: 2, pclk_manual: 1, pclk_div: 4
I (4995) ov5640: Calculated XVCLK: 20000000 Hz, REFIN: 10000000 Hz, VCO: 1800000000 Hz, PLL_CLK: 180000000 Hz, SYSCLK: 45000000 Hz, PCLK: 11250000 Hz
I (5065) CAM: OV5640 camera module detected
I (7065) MAIN: Camera info: PID: 0x5640, VER: 0x00, MIDL: 0x00, MIDH: 0x00

jasaw commented:
Interesting... `s->set_res_raw(s, xOffset + 1, yOffset + 1, xOffset + xTotal + 1, yOffset + yTotal + 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning);` gives me a red tint over the image.

Any good ideas on how we can narrow down the difference between your camera and mine? How should we handle this?


This is my OV5640 camera:
```
I (3854) camera: Detected OV5640 camera
I (4544) cam_hal: cam config ok
I (4564) ov5640: Set PLL: bypass: 0, multiplier: 180, sys_div: 4, pre_div: 2, root_2x: 0, pclk_root_div: 2, pclk_manual: 1, pclk_div: 4
I (4564) ov5640: Calculated XVCLK: 20000000 Hz, REFIN: 10000000 Hz, VCO: 1800000000 Hz, PLL_CLK: 180000000 Hz, SYSCLK: 45000000 Hz, PCLK: 11250000 Hz
I (4604) CAM: OV5640 camera module detected
I (6604) MAIN: Camera info: PID: 0x5640, VER: 0x00, MIDL: 0x00, MIDH: 0x00
```
![ov5640_closeup](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/60793990-ce7a-4b6e-b780-78fda8fd73e1)

SybexX commented:
I've tried a few things, but as soon as the values ​​are divisible by 2 without a remainder, the colors are swapped.

jasaw commented:
@SybexX What's your Espressif SDK version? 

Mine is this:
```
platformio run --environment esp32cam
Processing esp32cam (board: esp32cam; framework: espidf; platform: platformio/espressif32 @ 6.5.0)
--------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32cam.html
PLATFORM: Espressif 32 (6.5.0) > AI Thinker ESP32-CAM
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-espidf @ 3.50102.240122 (5.1.2)
 - tool-cmake @ 3.16.4
 - tool-esptoolpy @ 1.40501.0 (4.5.1)
 - tool-ninja @ 1.7.1
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0)
 - toolchain-xtensa-esp32 @ 12.2.0+20230208
Reading CMake configuration...

```

SybexX commented:
Processing esp32cam (board: esp32cam; framework: espidf; platform: platformio/espressif32 @ 6.5.0)
--------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32cam.html
PLATFORM: Espressif 32 (6.5.0) > AI Thinker ESP32-CAM
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-espidf @ 3.50102.240122 (5.1.2)
 - tool-cmake @ 3.16.4
 - tool-esptoolpy @ 1.40501.0 (4.5.1)
 - tool-idf @ 1.0.1
 - tool-mconf @ 1.4060000.20190628 (406.0.0)
 - tool-ninja @ 1.9.0
 - tool-riscv32-esp-elf-gdb @ 11.2.0+20220823
 - tool-xtensa-esp-elf-gdb @ 11.2.0+20230208
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0)
 - toolchain-xtensa-esp32 @ 12.2.0+20230208
Reading CMake configuration...

jasaw commented:
@SybexX I'm not sure what else we can do to track down the root cause of your camera red blue colour swap.
I have 2 hypothesis:
1. Your camera is actually different from mine, either different hardware revision or different firmware.
2. Your development environment is different from mine causing slightly different behaviour in the compiled firmware.

I have shared my firmware here for you to test to help isolate the problem.
This firmware also reads register 0x302A (Chip Revision) and the datasheet has 2 possible values for the process (0xA or 0xB), but my camera reports process value of zero.
`CAM: OV5640 camera module detected. Chip revision 0, process 0`

Other than this, I'm not sure what else we can do to track this issue down, then the next step is to probably have a red blue colour swap workaround flag that can be controlled via the web interface?

[AI-on-the-edge-device__update__debug_20240517.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15342429/AI-on-the-edge-device__update__debug_20240517.zip)

SybexX commented:
me too^^
I (4594) CAM: OV5640 camera module detected. Chip revision 0, process 0

I assume it's due to the camera's firmware, your number has an 8 and mine has a B.
DC-5M21-5640-8-03-V1
DC-5M21-5640-B-03-V1

Since there are no differences when querying the version and revision, the only way to do this is via the settings page.

maybe like this: [AI-on-the-edge-device-ov5640_support_changes.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15358653/AI-on-the-edge-device-ov5640_support_changes.zip)

jasaw commented:
@SybexX I've pushed your changes, however I did not comment out the code that limits the jpeg quality range. Was it commented out on purpose?
```C
    if (CCstatus.CamSensor_id == OV5640_PID)
    {
        qual = min(63, max(18, qual)); // Limit quality from 18..63 (values lower than 20 tent to be unstable)
    }
    else
    {
        qual = min(63, max(8, qual)); // Limit quality from 8..63 (values lower than 8 tent to be unstable)
    }
```

SybexX commented:
We already query the validity of the value in MainFlowControl.cpp and ClassFlowTakeImage.cpp and my OV5640 works with a value of 8 without any problems(even in poor lighting conditions), so I commented it out.
For the OV2640 you don't actually need a limit, for the OV5640 maybe like this:
´´´´´´
            case OV5640_PID:
                if (CCstatus.isImageColorSwaped == true)
                {
					qual = min(63, max(8, qual)); // Limit quality from 18..63 (values lower than 20 tent to be unstable)
                }
                else
                {
					qual = min(63, max(18, qual)); // Limit quality from 18..63 (values lower than 20 tent to be unstable)
                }			

                frameSizeX = 2592;
                frameSizeY = 1944;
                // max imageSize = ((frameSizeX - CCstatus.ImageWidth) / 8 / 4) - 1
                // 59 = ((2560 - 640) / 8 / 4) - 1
                if (imageSize < 59)
                {
                    _imageSize_temp = (59 - imageSize);
                }
                SanitizeZoomParams(_imageSize_temp, frameSizeX, frameSizeY, _imageWidth, _imageHeight, _offsetx, _offsety);
                SetCamWindow(s, frameSizeX, frameSizeY, _offsetx, _offsety, _imageWidth, _imageHeight, CCstatus.ImageWidth, CCstatus.ImageHeight);
                break;
´´´´´´

jasaw commented:
@SybexX OK, I have updated the jpeg limits for our 2 variants of OV5640 and removed the OV2640 jpeg limit. Do you know whether OV3660 has a limit? I guess we just update the OV3660 limit when someone can actually test it? For now we just handle it the same as OV2640?

SybexX commented:
So far I don't know anyone who has the OV3660, and if anyone does, to my knowledge there have never been any complaints about problems with the camera.
Therefore it probably doesn't matter whether it is treated like the OV5640 or OV2640^^

jasaw commented:
@SybexX Is there anymore changes or testing that we want to do on this PR before we merge it in?

SybexX commented:
Everything works as it should with my camera, so I don't have any other changes that need to be made.
Maybe it would be an advantage if some people with an OV5640 would test it first and report here, who knows how many variants of the camera there are ^^

SybexX commented:
I think I have found the main problem of swapping the colors.
In my tests I always had "Flip Image = true", when I deactivate it the color swap is no longer there.
Please test whether this is the case for you too.

SybexX commented:
I've tried it like this now:
```C
void CCamera::SetCamWindow(sensor_t *s, int frameSizeX, int frameSizeY, int xOffset, int yOffset, int xTotal, int yTotal, int xOutput, int yOutput)
{
    if (CCstatus.CamSensor_id == OV2640_PID)
    {
        s->set_res_raw(s, 0, 0, 0, 0, xOffset, yOffset, xTotal, yTotal, xOutput, yOutput, false, false);
    }
    else
    {
        // for CAMERA_OV5640 and CAMERA_OV3660
        bool scale = !(xOutput == xTotal && yOutput == yTotal);
        bool binning = (xTotal >= (frameSizeX >> 1));

        if (CCstatus.ImageVflip == true)
        {
            s->set_res_raw(s, xOffset, yOffset, xOffset + xTotal - 1, yOffset + yTotal - 1, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning);
        }
        else
        {
            s->set_res_raw(s, xOffset, yOffset, xOffset + xTotal, yOffset + yTotal, 0, 0, frameSizeX, frameSizeY, xOutput, yOutput, scale, binning);
        }
    }
}
```

Now is "xOffset + xTotal - 1" instead of "xOffset + xTotal + 1", because a strange line appeared in the image at "+"

jasaw commented:
@SybexX I have 2 different lots of OV5640 camera modules and both behave the same way as yours where vertical flip causes the image go turn red. I suggest we get rid of vertical flip altogether because vertical flip can be achieved by mirror + 180 degrees rotation.

SybexX commented:
I would reluctantly waive Vflip because the Hmirror function then doesn't make much sense anymore.
Some users do not have the ability to rotate the camera properly and have to rotate the image by 180°.
If these options are no longer available, you have to use rotation and this is handled by the ESP.
This will likely use more memory and put additional strain on the ESP, this increasing the processing time and possibly the temperature.

jasaw commented:
@SybexX I've removed the colour swap config and implemented the workaround based on vflip. I'm not sure why the build on github action is failing.

SybexX commented:
components/jomjol_controlcamera/ClassControllCamera.cpp:468:6: error: no declaration matches 'void CCamera::SetQualityZoomSize(int, framesize_t, bool, int, int, int)'
  468 | void CCamera::SetQualityZoomSize(int qual, framesize_t resol, bool zoomEnabled, int zoomOffsetX, int zoomOffsetY, int imageSize)
      |      ^~~~~~~
In file included from components/jomjol_controlcamera/ClassControllCamera.cpp:1:
components/jomjol_controlcamera/ClassControllCamera.h:95:10: note: candidate is: 'void CCamera::SetQualityZoomSize(int, framesize_t, bool, int, int, int, int)'
   95 |     void SetQualityZoomSize(int qual, framesize_t resol, bool zoomEnabled, int zoomOffsetX, int zoomOffsetY, int imageSize, int imageVflip);
      |          ^~~~~~~~~~~~~~~~~~
components/jomjol_controlcamera/ClassControllCamera.h:72:7: note: 'class CCamera' defined here
   72 | class CCamera
      |       ^~~~~~~

SybexX commented:
in components/jomjol_controlcamera/ClassControllCamera.cpp
change from:
void CCamera::SetQualityZoomSize(int qual, framesize_t resol, bool zoomEnabled, int zoomOffsetX, int zoomOffsetY, int imageSize)
to:
void CCamera::SetQualityZoomSize(int qual, framesize_t resol, bool zoomEnabled, int zoomOffsetX, int zoomOffsetY, int imageSize, int imageVflip);

jasaw commented:
Oops, I missed that line in my previous commit. 

I just received my autofocus OV5640 and I have many autofocus experimental changes.
![20240522_125703](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/4d3ef8dd-be41-4139-8c75-687b8f51e452)

Slider0007 commented:
Just a small side note: To my knowledge OV3660 is not supported up to now. It is not even compiled in the latest rolling.

https://github.com/jomjol/AI-on-the-edge-device/blob/rolling/code%2Fsdkconfig.defaults#L137

SybexX commented:
@jasaw Can you try out whether the following changes eliminate the red coloring when you have vertical flip activated?
Does your OV5640 with auto focus also have the same problems with vertical flip?

change in ClassControllCamera.cpp;
```C
esp_err_t CCamera::setSensorDatenFromCCstatus(void)
{
    sensor_t *s = esp_camera_sensor_get();

    if (s != NULL)
    {
        s->set_framesize(s, CCstatus.ImageFrameSize);
        s->set_quality(s, CCstatus.ImageQuality); 							// 0 - 63

        s->set_brightness(s, CCstatus.ImageBrightness); 						// -2 to 2
        s->set_contrast(s, CCstatus.ImageContrast);     						// -2 to 2
        s->set_saturation(s, CCstatus.ImageSaturation); 						// -2 to 2
        // s->set_sharpness(s, CCstatus.ImageSharpness);   						// auto-sharpness is not officially supported, default to 0
        SetCamSharpness(CCstatus.ImageAutoSharpness, CCstatus.ImageSharpness);

        s->set_denoise(s, CCstatus.ImageDenoiseLevel); 						// The OV2640 does not support it, OV3660 and OV5640 (0 to 8)

        s->set_special_effect(s, CCstatus.ImageSpecialEffect); 					// 0 to 6 (0 - No Effect, 1 - Negative, 2 - Grayscale, 3 - Red Tint, 4 - Green Tint, 5 - Blue Tint, 6 - Sepia)
        s->set_wb_mode(s, CCstatus.ImageWbMode);               					// 0 to 4 - if awb_gain enabled (0 - Auto, 1 - Sunny, 2 - Cloudy, 3 - Office, 4 - Home)

        s->set_ae_level(s, CCstatus.ImageAeLevel);        						// -2 to 2
        s->set_aec_value(s, CCstatus.ImageAecValue);      						// 0 to 1200
        s->set_agc_gain(s, CCstatus.ImageAgcGain);        						// 0 to 30
        s->set_gainceiling(s, CCstatus.ImageGainceiling); 						// Image gain (GAINCEILING_x2, x4, x8, x16, x32, x64 or x128)

        s->set_lenc(s, CCstatus.ImageLenc);         							// 0 = disable , 1 = enable
        s->set_gain_ctrl(s, CCstatus.ImageAgc);     							// 0 = disable , 1 = enable
        s->set_exposure_ctrl(s, CCstatus.ImageAec); 							// 0 = disable , 1 = enable

        s->set_hmirror(s, CCstatus.ImageHmirror); 							// 0 = disable , 1 = enable
        s->set_vflip(s, CCstatus.ImageVflip);     								// 0 = disable , 1 = enable
        s->set_aec2(s, CCstatus.ImageAec2);       								// 0 = disable , 1 = enable

        s->set_bpc(s, CCstatus.ImageBpc); 									// 0 = disable , 1 = enable
        s->set_wpc(s, CCstatus.ImageWpc); 								// 0 = disable , 1 = enable

        s->set_raw_gma(s, CCstatus.ImageRawGma); 							// 0 = disable , 1 = enable

        s->set_awb_gain(s, CCstatus.ImageAwbGain); 							// 0 = disable , 1 = enable
        s->set_whitebal(s, CCstatus.ImageAwb);     							// 0 = disable , 1 = enable

        s->set_dcw(s, CCstatus.ImageDcw); 									// 0 = disable , 1 = enable

        TickType_t xDelay2 = 1000 / portTICK_PERIOD_MS;
        vTaskDelay(xDelay2);

        return ESP_OK;
    }
    else
    {
        return ESP_FAIL;
    }
}
```

change in MainFlowControl.cpp;
```C
esp_err_t setCFstatusToCam(void)
{
    sensor_t *s = esp_camera_sensor_get();

    if (s != NULL)
    {
        s->set_framesize(s, CFstatus.ImageFrameSize);
        s->set_quality(s, CFstatus.ImageQuality); 							// 0 - 63

        s->set_brightness(s, CFstatus.ImageBrightness); 						// -2 to 2
        s->set_contrast(s, CFstatus.ImageContrast);     						// -2 to 2
        s->set_saturation(s, CFstatus.ImageSaturation); 						// -2 to 2
        // s->set_sharpness(s, CFstatus.ImageSharpness);   						// auto-sharpness is not officially supported, default to 0
        Camera.SetCamSharpness(CFstatus.ImageAutoSharpness, CFstatus.ImageSharpness);

        s->set_denoise(s, CFstatus.ImageDenoiseLevel); 						// The OV2640 does not support it, OV3660 and OV5640 (0 to 8)

        s->set_special_effect(s, CFstatus.ImageSpecialEffect); 					// 0 to 6 (0 - No Effect, 1 - Negative, 2 - Grayscale, 3 - Red Tint, 4 - Green Tint, 5 - Blue Tint, 6 - Sepia)
        s->set_wb_mode(s, CFstatus.ImageWbMode);               					// 0 to 4 - if awb_gain enabled (0 - Auto, 1 - Sunny, 2 - Cloudy, 3 - Office, 4 - Home)

        s->set_ae_level(s, CFstatus.ImageAeLevel);   							// -2 to 2
        s->set_aec_value(s, CFstatus.ImageAecValue); 							// 0 to 1200
        s->set_agc_gain(s, CFstatus.ImageAgcGain);   							// 0 to 30
        s->set_gainceiling(s, CFstatus.ImageGainceiling);

        s->set_lenc(s, CFstatus.ImageLenc);         								// 0 = disable , 1 = enable
        s->set_gain_ctrl(s, CFstatus.ImageAgc);     							// 0 = disable , 1 = enable
        s->set_exposure_ctrl(s, CFstatus.ImageAec); 							// 0 = disable , 1 = enable

        s->set_hmirror(s, CFstatus.ImageHmirror); 							// 0 = disable , 1 = enable
        s->set_vflip(s, CFstatus.ImageVflip);     								// 0 = disable , 1 = enable
        s->set_aec2(s, CFstatus.ImageAec2);       								// 0 = disable , 1 = enable

        s->set_bpc(s, CFstatus.ImageBpc); 									// 0 = disable , 1 = enable
        s->set_wpc(s, CFstatus.ImageWpc); 									// 0 = disable , 1 = enable

        s->set_raw_gma(s, CFstatus.ImageRawGma); 							// 0 = disable , 1 = enable

        s->set_awb_gain(s, CFstatus.ImageAwbGain); 							// 0 = disable , 1 = enable
        s->set_whitebal(s, CFstatus.ImageAwb);     							// 0 = disable , 1 = enable

        s->set_dcw(s, CFstatus.ImageDcw); 									// 0 = disable , 1 = enable

        TickType_t xDelay2 = 100 / portTICK_PERIOD_MS;
        vTaskDelay(xDelay2);

        return ESP_OK;
    }
    else
    {
        return ESP_FAIL;
    }
}
```

SybexX commented:
@jasaw I changed something regarding gainceiling because it didn't work properly with the OV5640
[AI-on-the-edge-device-ov5640_support.zip](https://github.com/jomjol/AI-on-the-edge-device/files/15446255/AI-on-the-edge-device-ov5640_support.zip)

jasaw commented:
@SybexX 
>Can you try out whether the following changes eliminate the red coloring when you have vertical flip activated?
Does your OV5640 with auto focus also have the same problems with vertical flip?

The config order changes does not fix the vertical flip red colour problem for me. Both my normal OV5640 and OV5640 autofocus behave the same way. From I can tell, the autofocus version behaves exactly the same way as the normal OV5640 in every way, except it has the autofocus VCM circuit attached at the side.

>I changed something regarding gainceiling because it didn't work properly with the OV5640

Good catch. Gain ceiling is a little hard to test because it only kicks in after it has reached maximum exposure level and the datasheet is not exactly clear that the gain ceiling registers take real value rather than enum. The espressif library gets it wrong, so would be nice to upstream your fix.

SybexX commented:
@jasaw It might be an advantage if you put something like this in the AI-on-the-edge-device/param-docs/parameter-pages/TakeImage/CamVflip.md:

!!! On some OV5640 Cameras, the image becomes reddish when Vflip is used in conjunction with the zoom function!

To avoid countless inquiries/reports.

jasaw commented:
@SybexX I've updated the vflip param page and resolved merge conflict.

jasaw commented:
I've been running my OV5640 for some time now. It is working well for me. Can we merge this in?

caco3 commented:
@SybexX Is this something we should add to the official firmware?
@jasaw  can you please fix the conflicts?

SybexX commented:
@caco3 yes, I would like to have this in there too.
It works as it should, does not consume any additional resources compared to the ov2640 and expands the supported hardware.

jasaw commented:
I've resolved the merge conflicts, but haven't eyeballed the merged code line by line to make sure everything merged correctly. I haven't tested the firmware. I'll OTA upgrade to my OV5640 system tomorrow.

caco3 commented:
OK, that sounds great.
We should add this to the new release and mention it clearly as this a large enhancement -> readme, documentation, release notes, ...
@jasaw since you are a core user of it, please test it it works as needed.

SybexX commented:
I tested it with an ov2640 and ov5640 and I didn't notice any errors/problems.
Let's see what @jasaw says when he tests it.

jasaw commented:
I've been running this firmware for more than a day now and it's working as expected.

Pferdebockwurst commented:
Very nice to see some progress regarding the cameras. Can you recommend one of them? Is one of the new supported models better than the other one? Is autofocus important? Thanks a lot guys!

jasaw commented:
@Pferdebockwurst Auto focus is **NOT** supported yet and auto focus also requires some soldering on most ESP32Cam boards. If you don't solder wires, it's still safe to use OV5640 modules with auto focus. The auto focus circuitry is completely separate from the OV5640.
I bought my OV5640 with and without auto focus from 3 different sellers on AliExpress and they're all working fine for me, so maybe any one will work as long as it's advertised to work with ESP32Cam?

Slider0007 commented:
@jasaw, @SybexX 
I'm not sure if the issue with tinted image is solved already, but during my last activities I also faced the tinted image issue from time to time. Ensure that only an odd offset is provided, solved it for me - at least with my camera. I have only one OV5640 camera therefore I cannot say if this is transferable or the issue here is something different. Just let you know...

SybexX commented:
@Slider0007 Unfortunately the problem could not be solved.
We also noticed the issue with the offset, but it didn't help solve this problem.
What worked for me caused a problem for jasaw and vice versa.
Therefore I suspect that it has something to do with the camera's firmware.
Therefore, a note about this was added to the Tooltip.

![Unbenannt](https://github.com/user-attachments/assets/eb1caf1a-981b-46fe-bbca-d362aca06bf3)
